### PR TITLE
[feat]: [PIE-8416]: InputSet reconciliation to handle different repos for input set and pipeline

### PR DESCRIPTION
**Pull request by adivishy1**
## Describe your changes
Have added code for making the getYamlDiff code work for input sets and overlay input sets which are created in the different repo compared to the pipeline repo. Have tested out the reconciliation flows for:
1.  input sets in same repo and different repo as pipeline
2. overlay input sets in the same repo and different repo as the pipeline

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)

### CODE REVIEW

**Comment by mohitg0795** [comment]
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java` line 366 at r1](https://reviewable.io/reviews/harness/harness-core/44157#-NOxgBBw-WSDeb4d0ivj:-NOxgBBx30ugIVpS6e02:b2a4awp) ([raw file](https://github.com/harness/harness-core/blob/41b033f9dcb36d9544f81f61e6f4216120fb818e/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java#L366)):*
> ```Java
>     if (!inputSetsApiUtils.isSameRepoForPipelineAndInputSetsAccountSettingEnabled(accountId)
>         && EmptyPredicate.isEmpty(pipelineBranch)) {
>       throw new InvalidRequestException("Pipeline branch not provided.");
> ```

Lets create constant and use everywhere as a reference.

**Review by mohitg0795 — COMMENTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 9 file(s) to review. 0 of 2 approvals obtained.

**Review by adivishy1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 10 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java` line 366 at r1](https://reviewable.io/reviews/harness/harness-core/44157#-NOxgBBw-WSDeb4d0ivj:-NOxnXqH5SsSGiZxxuH9:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/41b033f9dcb36d9544f81f61e6f4216120fb818e/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java#L366)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Lets create constant and use everywhere as a reference.
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 9 of 9 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 10 file(s) to review. 1 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 7 of 9 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 1 change requests. 1 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java` line 203 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxwLAvBukvTOdyHsQc:-NOxwLAvBukvTOdyHsQd:b-flpdgm) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java#L203)):*
> ```Java
>   }
> 
>   public String getPipelineTemplate(String pipelineYaml, List<String> stageIdentifiers) {
> ```

add comment if pipeline yaml is there prefer this, this is for devs only
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java` line 148 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxvIu10IlFqUMxh5is:-NOxvIu10IlFqUMxh5it:bgxdxox) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java#L148)):*
> ```Java
>   }
> 
>   public InputSetYamlDiffDTO getYAMLDiff(GitSyncSdkService gitSyncSdkService, PMSInputSetService inputSetService,
> ```

break into different methods
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java` line 157 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxuKCq8luBm44O9TKf:-NOxuKCq8luBm44O9TKg:b-7hmp0c) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java#L157)):*
> ```Java
>         accountId, orgIdentifier, projectIdentifier, pipelineIdentifier, inputSetIdentifier, false, false, true);
>     if (optionalInputSetMetadataEntity.isEmpty()) {
>       throw new InvalidRequestException(
> ```

move to service layer with hint and explanation 



_Code quote:_
```Java
    Optional<InputSetEntity> optionalInputSetMetadataEntity = inputSetService.getMetadataWithoutValidations(
        accountId, orgIdentifier, projectIdentifier, pipelineIdentifier, inputSetIdentifier, false, false, true);
    if (optionalInputSetMetadataEntity.isEmpty()) {
      throw new InvalidRequestException(
          String.format("InputSet with the given ID: %s does not exist or has been deleted", inputSetIdentifier));
    }
```
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java` line 180 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxuxgBBVntdD05NCBG:-NOxuxgBBVntdD05NCBH:b-3mh7nh) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java#L180)):*
> ```Java
>       }
>     } else {
>       optionalInputSetEntity = inputSetService.getWithoutValidations(
> ```

add comments to say its different repo
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java` line 197 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxvD4E2fUZ-qzh3z5i:-NOxvD4E2fUZ-qzh3z5j:b-gtho5z) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java#L197)):*
> ```Java
>           pipelineIdentifier, pipelineBranch, pipelineRepoID);
>     } else {
>       PipelineEntity pipelineEntity;
> ```

we should add comments to know if its old or new git exp
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/OverlayInputSetValidationHelper.java` line 168 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxvhvO0Om4SbVekCIL:-NOxvhvO0Om4SbVekCIM:bqmtqvk) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/OverlayInputSetValidationHelper.java#L168)):*
> ```Java
>       }
>       inputSets.add(inputSetService.getWithoutValidations(
>           accountId, orgIdentifier, projectIdentifier, pipelineIdentifier, identifier, false, true));
> ```

verify once all callers do not require yaml

**Review by archit-harness — APPROVED** [review]
> Reviewed 5 of 5 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 7 file(s) to review. 1 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java` line 148 at r2](https://reviewable.io/reviews/harness/harness-core/44157#-NOxvIu10IlFqUMxh5is:-NOyZtot26tynvO7UOdC:b-t7xo89) ([raw file](https://github.com/harness/harness-core/blob/bff7e5e42501b28c3b872904ff42c59c4b485bb7/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/service/InputSetValidationHelper.java#L148)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

we should break this into further methods

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: 7 file(s) to review. 1 of 2 approvals obtained.

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 1 of 9 files at r1, 2 of 3 files at r2, 4 of 5 files at r3, 1 of 2 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44157)* status: :shipit: complete! ✓

File: pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java [hunk 1]
```diff
@@ -210,9 +221,12 @@ public String getMergedYamlFromInputSetReferencesAndRuntimeInputYaml(String acco
       String projectIdentifier, String pipelineIdentifier, List<String> inputSetReferences, String pipelineBranch,
       String pipelineRepoID, List<String> stageIdentifiers, String lastYamlToMerge) {
     Set<String> inputSetVersions = new HashSet<>();
-    PipelineGitXHelper.setupGitParentEntityDetails(accountId, orgIdentifier, projectIdentifier);
-    PipelineEntity pipelineEntity = getPipelineEntity(
-        accountId, orgIdentifier, projectIdentifier, pipelineIdentifier, pipelineBranch, pipelineRepoID, false);
+    GitSyncBranchContext branchContext = setupGitContext(accountId, orgIdentifier, projectIdentifier, pipelineBranch);
```
Comment (vcalasansh) [inline pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java:224]: This

**Pull request merged**